gnomad/utils/constraint.py

```diff
@@ -0,0 +1,162 @@
+# noqa: D100
+# cSpell: disable
+from typing import Any, Optional, Tuple, Union
+import logging
+
+import hail as hl
+
+logging.basicConfig(
+    format="%(asctime)s (%(name)s %(lineno)s): %(message)s",
+    datefmt="%m/%d/%Y %I:%M:%S %p",
+)
+logger = logging.getLogger("constraint_pipeline")
+logger.setLevel(logging.INFO)
+
+
+def annotate_with_mu(
+    ht: hl.Table,
+    mutation_ht: hl.Table,
+    output_col: str = "mu_snp",
+) -> hl.Table:
+    """
+    Annotate SNP mutation rate for the input Table.
+
+    :param ht: Input Table.
+    :param mutation_ht: Mutation rate Table.
+    :param output_col: Name for mutational rate annotation. Defaults to 'mu_snp'.
+    :return: Table with mutational rate annotation added (default name for annotation is 'mu_snp').
+    """
+    keys = mutation_ht.key
+    mu = hl.literal(
+        mutation_ht.aggregate(
+            hl.dict(
+                hl.agg.collect(
+                    (hl.struct(**{k: mutation_ht[k] for k in keys}), mutation_ht.mu_snp)
+                )
+            )
+        )
+    )
+    mu = mu.get(hl.struct(**{k: ht[k] for k in keys}))
+    return ht.annotate(
+        **{output_col: hl.case().when(hl.is_defined(mu), mu).or_error("Missing mu")}
+    )
+
+
+def count_variants(
+    ht: hl.Table,
+    count_singletons: bool = False,
+    count_downsamplings: Optional[Tuple[str]] = (),
+    additional_grouping: Optional[Tuple[str]] = (),
+    partition_hint: int = 100,
+    omit_methylation: bool = False,
+    use_table_group_by: bool = False,
+    singleton_expr: hl.expr.BooleanExpression = None,
+    max_af: Optional[float] = None,
+    freq_expr: hl.expr.ArrayExpression = None,
+    freq_meta_expr: hl.expr.ArrayExpression = None,
+) -> Union[hl.Table, Any]:
+    """
+    Count number of observed or possible variants by context, ref, alt, and optionally methylation_level.
+
+    :param ht: Input Hail Table.
+    :param count_singletons: Whether to count singletons. Defaults to False.
+    :param count_downsamplings: List of populations to use for downsampling counts. Defaults to ().
+    :param additional_grouping: Additional features to group by. i.e. exome_coverage. Defaults to ().
+    :param partition_hint: Target number of partitions for aggregation. Defaults to 100.
+    :param omit_methylation: Whether to omit 'methylation_level' from the grouping when counting variants. Defaults to False.
+    :param use_table_group_by: Whether to force grouping. Defaults to False.
```

Comment:
It's really just how to return the aggregate count info. I think in the description above you should have an example.